File: DevExpress-Demo/CS/PropertyGridDemo.Wpf/Data/DynamicallyAssignDataEditorsData.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Media;
using DevExpress.Xpf.Editors;
using System.ComponentModel.DataAnnotations;
using DevExpress.Mvvm.DataAnnotations;

namespace PropertyGridDemo {
    [MetadataType(typeof(DynamicallyAssignDataEditorsMetadata))]
    public class DynamicallyAssignDataEditorsData {
        const string SimpleEditorsGroup = "Simple editors";
        const string PopupEditorsGroup = "Popup editors";
        const string MiscEditorsGroup = "Misc editors";

        public DynamicallyAssignDataEditorsData() {
            var list = new MultiEditorsList();
            foreach(PropertyDescriptor property in TypeDescriptor.GetProperties(this)) {
                property.SetValue(this, list.GetValue(property.Name));
            }
        }

        [Display(Name = "TextEdit", GroupName = SimpleEditorsGroup)]
        public string Name { get; set; }

        [Display(Name = "TextEdit (with numeric mask)", Description = "NumericTextEdit", GroupName = SimpleEditorsGroup)]
        public long ID { get; set; }

        [Display(Name = "TextEdit (with RegEx mask)", Description = "RegExTextEdit", GroupName = SimpleEditorsGroup)]
        public string Code { get; set; }

        [Display(Name = "PasswordBoxEdit", GroupName = SimpleEditorsGroup)]
        public string Password { get; set; }

        [Display(Name = "ButtonEdit", GroupName = SimpleEditorsGroup)]
        public string Countries { get; set; }


        [Display(Name = "ComboBoxEdit (with AutoComplete)", Description = "AutoCompleteComboBoxEdit", GroupName = PopupEditorsGroup)]
        public string Country { get; set; }

        [Display(Name = "SearchLookUpEdit", GroupName = PopupEditorsGroup)]
        public long Category1 { get; set; }

        [Display(Name = "LookUpEdit", GroupName = PopupEditorsGroup)]
        public long Category2 { get; set; }

        [Display(Name = "MemoEdit", GroupName = PopupEditorsGroup)]
        public string Notes { get; set; }

        [Display(Name = "DateEdit", GroupName = PopupEditorsGroup)]
        public DateTime Date1 { get; set; }

        [Display(Name = "DatePicker", GroupName = PopupEditorsGroup)]
        public DateTime Date2 { get; set; }

        [Display(Name = "FontEdit", GroupName = PopupEditorsGroup)]
        public FontFamily Font { get; set; }

        [Display(Name = "PopupCalcEdit", GroupName = PopupEditorsGroup)]
        public decimal UnitPrice { get; set; }

        [Display(Name = "PopupColorEdit", GroupName = PopupEditorsGroup)]
        public Color Color { get; set; }

        [Display(Name = "PopupBrushEdit", GroupName = PopupEditorsGroup)]
        public Brush Brush { get; set; }

        [Display(Name = "PopupImageEdit", GroupName = PopupEditorsGroup)]
        public ImageSource Picture { get; set; }


        [Display(Name = "TrackBarEdit", GroupName = MiscEditorsGroup)]
        public double UnitsInStock { get; set; }

        [Display(Name = "ZoomTrackBarEdit", GroupName = MiscEditorsGroup)]
        public double ReorderLevel { get; set; }

        [Display(Name = "RangeTrackBarEdit", GroupName = MiscEditorsGroup)]
        public TrackBarEditRange Range { get; set; }

        [Display(Name = "ProgressBarEdit", GroupName = MiscEditorsGroup)]
        public double UnitsOnOrder { get; set; }

        [Display(Name = "CheckEdit", GroupName = MiscEditorsGroup)]
        public bool Discontinued { get; set; }

        [Display(Name = "ToggleSwitchEdit", GroupName = MiscEditorsGroup)]
        public bool Discontinued2 { get; set; }

        [Display(Name = "SpinEdit", GroupName = MiscEditorsGroup)]
        public double Discount { get; set; }

        [Display(Name = "ListBoxEdit", GroupName = MiscEditorsGroup)]
        public string PalleteSize { get; set; }

        [Display(Name = "BarCodeEdit", GroupName = MiscEditorsGroup)]
        public int Data { get; set; }

        [Display(Name = "RatingEdit", GroupName = MiscEditorsGroup)]
        public double Rating { get; set; }

        [Display(Name = "HyperlinkEdit", GroupName = MiscEditorsGroup)]
        public string HyperLink { get; set; }
    }
    public static class DynamicallyAssignDataEditorsMetadata {
        public static void BuildMetadata(MetadataBuilder<DynamicallyAssignDataEditorsData> builder) {
            builder.Property(x => x.Password)
                .PasswordDataType();
            builder.Property(x => x.Notes)
                .MultilineTextDataType();
        }
    }
}

File: DevExpress-Demo/VB/EditorsDemo.Wpf/Data/DynamicallyAssignDataEditorsData.vb
Imports System.ComponentModel
Imports System.Windows.Media
Imports DevExpress.Xpf.Editors
Imports System.ComponentModel.DataAnnotations
Imports DevExpress.Mvvm.DataAnnotations

Namespace EditorsDemo

    <MetadataType(GetType(DynamicallyAssignDataEditorsMetadata))>
    Public Class DynamicallyAssignDataEditorsData

        Const SimpleEditorsGroup As String = "Simple editors"

        Const PopupEditorsGroup As String = "Popup editors"

        Const MiscEditorsGroup As String = "Misc editors"

        Public Sub New()
            Dim list = New PropertyGridMultiEditorsList()
            For Each [property] As PropertyDescriptor In TypeDescriptor.GetProperties(Me)
                [property].SetValue(Me, list.GetValue([property].Name))
            Next
        End Sub

        <Display(Name:="TextEdit", GroupName:=SimpleEditorsGroup)>
        Public Property Name As String

        <Display(Name:="TextEdit (with numeric mask)", Description:="NumericTextEdit", GroupName:=SimpleEditorsGroup)>
        Public Property ID As Long

        <Display(Name:="TextEdit (with RegEx mask)", Description:="RegExTextEdit", GroupName:=SimpleEditorsGroup)>
        Public Property Code As String

        <Display(Name:="PasswordBoxEdit", GroupName:=SimpleEditorsGroup)>
        Public Property Password As String

        <Display(Name:="ButtonEdit", GroupName:=SimpleEditorsGroup)>
        Public Property Countries As String

        <Display(Name:="AutoSuggestEdit", GroupName:=PopupEditorsGroup)>
        Public Property Product As String

        <Display(Name:="ComboBoxEdit (with AutoComplete)", Description:="AutoCompleteComboBoxEdit", GroupName:=PopupEditorsGroup)>
        Public Property Country As String

        <Display(Name:="SearchLookUpEdit", GroupName:=PopupEditorsGroup)>
        Public Property Category1 As Long

        <Display(Name:="LookUpEdit", GroupName:=PopupEditorsGroup)>
        Public Property Category2 As Long

        <Display(Name:="MemoEdit", GroupName:=PopupEditorsGroup)>
        Public Property Notes As String

        <Display(Name:="DateEdit", GroupName:=PopupEditorsGroup)>
        Public Property Date1 As Date

        <Display(Name:="DatePicker", GroupName:=PopupEditorsGroup)>
        Public Property Date2 As Date

        <Display(Name:="FontEdit", GroupName:=PopupEditorsGroup)>
        Public Property Font As FontFamily

        <Display(Name:="PopupCalcEdit", GroupName:=PopupEditorsGroup)>
        Public Property UnitPrice As Decimal

        <Display(Name:="PopupColorEdit", GroupName:=PopupEditorsGroup)>
        Public Property Color As Color

        <Display(Name:="PopupBrushEdit", GroupName:=PopupEditorsGroup)>
        Public Property Brush As Brush

        <Display(Name:="PopupImageEdit", GroupName:=PopupEditorsGroup)>
        Public Property Picture As ImageSource

        <Display(Name:="TrackBarEdit", GroupName:=MiscEditorsGroup)>
        Public Property UnitsInStock As Double

        <Display(Name:="ZoomTrackBarEdit", GroupName:=MiscEditorsGroup)>
        Public Property ReorderLevel As Double

        <Display(Name:="RangeTrackBarEdit", GroupName:=MiscEditorsGroup)>
        Public Property Range As TrackBarEditRange

        <Display(Name:="ProgressBarEdit", GroupName:=MiscEditorsGroup)>
        Public Property UnitsOnOrder As Double

        <Display(Name:="CheckEdit", GroupName:=MiscEditorsGroup)>
        Public Property Discontinued As Boolean

        <Display(Name:="ToggleSwitchEdit", GroupName:=MiscEditorsGroup)>
        Public Property Discontinued2 As Boolean

        <Display(Name:="SpinEdit", GroupName:=MiscEditorsGroup)>
        Public Property Discount As Double

        <Display(Name:="ListBoxEdit", GroupName:=MiscEditorsGroup)>
        Public Property PalleteSize As String

        <Display(Name:="BarCodeEdit", GroupName:=MiscEditorsGroup)>
        Public Property Data As Integer

        <Display(Name:="RatingEdit", GroupName:=MiscEditorsGroup)>
        Public Property Rating As Double

        <Display(Name:="HyperlinkEdit", GroupName:=MiscEditorsGroup)>
        Public Property HyperLink As String
    End Class

    Public Module DynamicallyAssignDataEditorsMetadata

        Public Sub BuildMetadata(ByVal builder As MetadataBuilder(Of DynamicallyAssignDataEditorsData))
            builder.[Property](Function(x) x.Password).PasswordDataType()
            builder.[Property](Function(x) x.Notes).MultilineTextDataType()
        End Sub
    End Module
End Namespace
